phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java

```diff
@@ -0,0 +1,266 @@
+ * for the specific language governing permissions and limitations under the License.
+ */
+package org.apache.phoenix.schema.json;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.Arrays;
+
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.exception.SQLExceptionCode;
+import org.apache.phoenix.exception.SQLExceptionInfo;
+import org.codehaus.jackson.JsonFactory;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.JsonParser;
+import org.codehaus.jackson.JsonParser.Feature;
+import org.codehaus.jackson.JsonProcessingException;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.codehaus.jackson.node.ValueNode;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * The {@link PhoenixJson} wraps json and uses Jackson library to parse and traverse the json. It
+ * should be used to represent the JSON data type and also should be used to parse Json data and
+ * read the value from it. It always conside the last value if same key exist more than once.
+ */
+public class PhoenixJson implements Comparable<PhoenixJson> {
+    private final JsonNode node;
+    private String jsonAsString;
+
+    /**
+     * Static Factory method to get an {@link PhoenixJson} object. It also validates the json and
+     * throws {@link SQLException} if it is invalid with line number and character.
+     * @param data Buffer that contains data to parse
+     * @param offset Offset of the first data byte within buffer
+     * @param length Length of contents to parse within buffer
+     * @return {@link PhoenixJson}.
+     * @throws SQLException
+     */
+    public static PhoenixJson getInstance(byte[] jsonData, int offset, int length)
+            throws SQLException {
+        String jsonDataStr = Bytes.toString(jsonData, offset, length);
+        return getInstance(jsonDataStr);
+    }
+
+    /**
+     * Static Factory method to get an {@link PhoenixJson} object. It also validates the json and
+     * throws {@link SQLException} if it is invalid with line number and character.
+     * @param jsonData Json data as {@link String}.
+     * @return {@link PhoenixJson}.
+     * @throws SQLException
+     */
+    public static PhoenixJson getInstance(String jsonData) throws SQLException {
+        try {
+            JsonFactory jsonFactory = new JsonFactory();
+            JsonParser jsonParser = jsonFactory.createJsonParser(jsonData);
+            PhoenixJson phoenixJson = getInstanceInternal(jsonParser);
+            /*
+             * input data has been stored as it is, since some data is lost when json parser runs,
+             * for example if a JSON object within the value contains the same key more than once
+             * then only last one is stored rest all of them are ignored, which will defy the
+             * contract of PJsonDataType of keeping user data as it is.
+             */
+            phoenixJson.setJsonAsString(jsonData);
```

Comment:
Pass the String through the PhoenixJson constructor. PhoenixJson should be immutable.